src/apps/chifra/pkg/cache/read.go

```diff
@@ -0,0 +1,813 @@
+	}
+	err = validateHeader(header)
+	if err != nil {
+		return
+	}
+
+	err = read(&block.GasLimit)
+	if err != nil {
+		return
+	}
+
+	err = read(&block.GasUsed)
+	if err != nil {
+		return
+	}
+
+	err = readHash(reader, &block.Hash)
+	if err != nil {
+		return
+	}
+
+	err = read(&block.BlockNumber)
+	if err != nil {
+		return
+	}
+
+	err = readHash(reader, &block.ParentHash)
+	if err != nil {
+		return
+	}
+
+	err = readAddress(reader, &block.Miner)
+	if err != nil {
+		return
+	}
+
+	err = read(&block.Difficulty)
+	if err != nil {
+		return
+	}
+
+	err = read(&block.Finalized)
+	if err != nil {
+		return
+	}
+
+	err = readTimestamp(read, &block.Timestamp)
+	if err != nil {
+		return
+	}
+
+	err = readBigUint(reader, &block.BaseFeePerGas)
+	if err != nil {
+		return
+	}
+
+	err = readFromArray(reader, &block.Transactions, ReadTransaction)
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+func ReadTransaction(reader *bufio.Reader) (tx *types.SimpleTransaction, err error) {
+	tx = &types.SimpleTransaction{}
+	read := createReadFn(reader)
+
+	header := &cacheHeader{}
+	err = readCacheHeader(reader, header)
+	if err != nil {
+		return
+	}
+	err = validateHeader(header)
+	if err != nil {
+		return
+	}
+
+	err = readHash(reader, &tx.Hash)
+	if err != nil {
+		return
+	}
+
+	err = readHash(reader, &tx.BlockHash)
+	if err != nil {
+		return
+	}
+
+	err = read(&tx.BlockNumber)
+	if err != nil {
+		return
+	}
+
+	err = read(&tx.TransactionIndex)
+	if err != nil {
+		return
+	}
+
+	err = read(&tx.Nonce)
+	if err != nil {
+		return
+	}
+
+	err = readTimestamp(read, &tx.Timestamp)
+	if err != nil {
+		return
+	}
+
+	err = readAddress(reader, &tx.From)
+	if err != nil {
+		return
+	}
+
+	err = readAddress(reader, &tx.To)
+	if err != nil {
+		return
+	}
+
+	err = readBigUint(reader, &tx.Value)
+	if err != nil {
+		return
+	}
+
+	err = readBigUint(reader, &tx.ExtraValue1)
+	if err != nil {
+		return
+	}
+
+	err = readBigUint(reader, &tx.ExtraValue2)
+	if err != nil {
+		return
+	}
+
+	err = read(&tx.Gas)
+	if err != nil {
+		return
+	}
+
+	err = read(&tx.GasPrice)
+	if err != nil {
+		return
+	}
+
+	err = read(&tx.MaxFeePerGas)
+	if err != nil {
+		return
+	}
+
+	err = read(&tx.MaxPriorityFeePerGas)
+	if err != nil {
+		return
+	}
+
+	err = readString(reader, &tx.Input)
+	if err != nil {
+		return
+	}
+
+	err = readUintAsBool(read, &tx.IsError)
+	if err != nil {
+		return
+	}
+
+	err = readUintAsBool(read, &tx.HasToken)
+	if err != nil {
+		return
+	}
+
+	err = read(&tx.Cachebits)
+	if err != nil {
+		return
+	}
+
+	err = read(&tx.Reserved2)
+	if err != nil {
+		return
+	}
+
+	receipt, err := ReadReceipt(reader)
+	if err != nil {
+		return
+	}
+	tx.Receipt = receipt
+
+	err = readFromArray(reader, &tx.Traces, ReadTrace)
+	if err != nil {
+		return
+	}
+
+	articulatedTx, err := ReadFunction(reader)
+	if err != nil {
+		return
+	}
+	tx.ArticulatedTx = articulatedTx
+
+	return
+}
+
+func ReadReceipt(reader *bufio.Reader) (receipt *types.SimpleReceipt, err error) {
```

Comment:
Eventually, we'll move this into the auto-generated type specific file. Please add a TODO, but don't move this yet.